homeassistant/components/slack/notify.py

```
@@ -144,6 +155,70 @@ def __init__(
         self._hass = hass
         self._client = client
         self._config = config
+        self._channel_id_cache: dict[str, str] = {}
+
+    async def _async_get_dm_channel_id(self, user_id: str) -> str | None:
+        """Get the DM channel ID for a user ID.
+
+        Opens a DM channel with the user if one doesn't exist.
+        """
+        try:
+            response = await self._client.conversations_open(users=[user_id])
+            if response["ok"]:
+                return cast(str, response["channel"]["id"])
+
+            _LOGGER.error(
+                "Failed to open DM with user %s: %s", user_id, response["error"]
+            )
+        except SlackApiError as err:
+            _LOGGER.error("Error opening DM with user %s: %r", user_id, err)
+
+        return None
+
+    async def _async_get_channel_id_by_name(self, channel_name: str) -> str | None:
+        """Get channel ID by name, checking cache first."""
+        if channel_id := self._channel_id_cache.get(channel_name):
+            return channel_id
+
+        try:
+            channels = []
+            for channel_type in ("public_channel", "private_channel"):
+                response = await self._client.conversations_list(types=channel_type)
+                if not response["ok"]:
+                    _LOGGER.error(
+                        "Error listing %s channels: %s",
+                        channel_type,
+                        response.get("error", "Unknown error"),
+                    )
```

Same, we need to raise instead of silently fail.